src/D2L.Security.OAuth2/Provisioning/Default/AuthServiceClient.cs

```diff
@@ -45,17 +46,60 @@ Uri authEndpoint
 		/// <param name="assertion">A JWT signed by the private key of the entity requesting the token</param>
 		/// <param name="scopes">List of scopes to include in the access token</param>
 		/// <returns>A JWT token from the auth service signed with the auth service's private key</returns>
+		/// <exception cref="AuthServiceException">
+		/// The auth service could not be reached, or it did not respond with
+		/// a status code indicating success.
+		/// </exception>
 		async Task<IAccessToken> IAuthServiceClient.ProvisionAccessTokenAsync(
 			string assertion,
 			IEnumerable<Scope> scopes
 		) {
 			string requestBody = BuildFormContents( assertion, scopes );
-			using( HttpResponseMessage response = await MakeRequest( requestBody ).SafeAsync() ) {
-				response.EnsureSuccessStatusCode();
-
-				using( var resultStream = await response.Content.ReadAsStreamAsync().SafeAsync() ) {
-					IAccessToken accessToken = SerializationHelper.ExtractAccessToken( resultStream );
-					return accessToken;
+			HttpResponseMessage response = null;
+			try {
+				try {
+					response = await MakeRequest( requestBody ).SafeAsync();
+				} catch( TaskCanceledException exception ) {
+					throw new AuthServiceException(
+						errorType: ServiceErrorType.Timeout,
+						message: "The web request to the Auth Service has timed out.",
+						innerException: exception
+					);
+				} catch( Exception exception ) {
+					throw new AuthServiceException(
+						errorType: ServiceErrorType.ConnectionFailure,
+						message: "Could not establish a connection with the Auth Service.",
+						innerException: exception
+					);
+				}
+				
+				if( !response.IsSuccessStatusCode ) {
+					throw new AuthServiceException(
+						errorType: ServiceErrorType.ErrorResponse,
+						serviceStatusCode: response.StatusCode,
+						message: string.Concat(
+							"The Auth Service did not respond with a status code indicating success. ",
+							"(Received ", ((int)response.StatusCode).ToString(), " ",
+							response.ReasonPhrase, ")"

```

Comment:
Here's (most of the relevant) bits of what I had, for reference. Think it's reasonable / could do something similar @mpharoah-d2l ?

```c#
using System.IO;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Json;

namespace D2L.Security.OAuth2.Provisioning.Default {
	[DataContract]
	internal sealed class AuthServiceErrorBody {

		[DataMember]
		public string error { get; set; }

		[DataMember]
		public string error_description { get; set; }

		internal static bool TryParse( Stream bodyStream, out AuthServiceErrorBody result ) {
			var serializer = new DataContractJsonSerializer( typeof( AuthServiceErrorBody ) );

			try {
				result = (AuthServiceErrorBody)serializer.ReadObject( bodyStream );
				return true;
			} catch {
				result = null;
				return false;
			}
		}
	}
}
```

```c#
internal static async Task<AuthServiceProvisioningException> MakeFromResponse( HttpResponseMessage response ) {
			string message;

			try {
				using( var bodyStream = await response.Content.ReadAsStreamAsync().SafeAsync() ) {
					AuthServiceErrorBody errorBody;
					if( AuthServiceErrorBody.TryParse( bodyStream, out errorBody ) ) {
						message = string.Format( CultureInfo.InvariantCulture, "{0}: {1}", errorBody.error, errorBody.error_description );
					} else {
						bodyStream.Seek( 0, SeekOrigin.Begin );
						using( var reader = new StreamReader( bodyStream ) ) {
							message = string.Format( CultureInfo.InvariantCulture, "{0}: {1}", response.ReasonPhrase, reader.ReadToEnd() );
						}
					}
				}
			} catch( Exception ) {
				message = response.ReasonPhrase;
			}

			return new AuthServiceProvisioningException(
				errorType: ServiceErrorType.ErrorResponse,
				proposedStatusCode: HttpStatusCode.InternalServerError,
				message: string.Format( CultureInfo.InvariantCulture, MESSAGE_FORMAT, message ),
				serviceStatusCode: response.StatusCode
			);
		}
```